# Official AUTO on mainline: DDR3 then LPDDR4, no WL, await spin-timeout.
#
--- a/arch/arm/mach-sunxi/dram_helpers.c
+++ b/arch/arm/mach-sunxi/dram_helpers.c
@@ -17,11 +17,14 @@
  */
 void mctl_await_completion(u32 *reg, u32 mask, u32 val)
 {
-	unsigned long tmo = timer_get_us() + 1000000;
+	u32 n = 0x400000;
 
+	/* Do not use timer_get_us(): SPL timer is unreliable before DRAM.
+	 * Spin then return so rank/type probe can fail instead of hanging.
+	 */
 	while ((readl(reg) & mask) != val) {
-		if (timer_get_us() > tmo)
-			panic("Timeout initialising DRAM\n");
+		if (!--n)
+			return;
 	}
 }
 
--- a/arch/arm/mach-sunxi/dram_sun50i_h616.c
+++ b/arch/arm/mach-sunxi/dram_sun50i_h616.c
@@ -1362,27 +1362,10 @@
 	return mctl_ctrl_init(para, config);
 }
 
-static bool walnutpi_any_phy;
-
-static bool walnutpi_sdram_works(void)
-{
-	writel(0x5a5aa5a5, CFG_SYS_SDRAM_BASE);
-	writel(0xa5a55a5a, CFG_SYS_SDRAM_BASE + 0x80);
-	dsb();
-	if (readl(CFG_SYS_SDRAM_BASE) != 0x5a5aa5a5)
-		return false;
-	if (readl(CFG_SYS_SDRAM_BASE + 0x80) != 0xa5a55a5a)
-		return false;
-	return true;
-}
-
 static bool walnutpi_rank_ok(const struct dram_para *para,
 			     struct dram_config *config)
 {
-	if (!mctl_core_init(para, config))
-		return false;
-	walnutpi_any_phy = true;
-	return walnutpi_sdram_works();
+	return mctl_core_init(para, config);
 }
 
 static bool mctl_auto_detect_rank_width(const struct dram_para *para,
@@ -1465,61 +1448,28 @@
 	return (1ULL << (config->cols + config->rows + 3)) * width * config->ranks;
 }
 
-static void walnutpi_dram_volt(unsigned int mv)
-{
-#ifdef CONFIG_AXP313_POWER
-	if (axp_set_dcdc3(mv))
-		printf(" v%u?", mv);
-	else
-		udelay(5000);
-#endif
-}
-
-/* OrangePi Zero2 (H616 DDR3) mainline: WL+cal+train, no DX bit-delay. */
-static void walnutpi_load_ddr3(struct dram_para *para, unsigned int clk)
+/* Official walnutpi AUTO: DDR3 uses read-cal + R/W training, no write
+ * leveling, no DX bit-delay. LPDDR4 keeps vendor TPR11/12 bit-delay.
+ */
+static void walnutpi_para_ddr3(struct dram_para *para)
 {
 	para->type = SUNXI_DRAM_TYPE_DDR3;
-	para->clk = clk;
-	para->dx_odt = 0x08080808;
-	para->tpr6 = 0x3300c080;
-	para->tpr10 = 0x00f83438;
+	para->clk = 792;
+	para->tpr6 = 0x4400c080;
+	para->tpr10 = TPR10_READ_CALIBRATION | TPR10_READ_TRAINING |
+		      TPR10_WRITE_TRAINING;
 	para->tpr11 = 0;
 	para->tpr12 = 0;
 }
 
-static void walnutpi_load_lpddr4(struct dram_para *para, u32 tpr6, u32 tpr10,
-				 u32 tpr11, u32 tpr12)
+static void walnutpi_para_lpddr4(struct dram_para *para)
 {
 	para->type = SUNXI_DRAM_TYPE_LPDDR4;
 	para->clk = 792;
-	para->dx_odt = 0x07070707;
-	para->tpr6 = tpr6;
-	para->tpr10 = tpr10;
-	para->tpr11 = tpr11;
-	para->tpr12 = tpr12;
-}
-
-static void walnutpi_load_ddr3_vendor(struct dram_para *para)
-{
-	para->type = SUNXI_DRAM_TYPE_DDR3;
-	para->clk = 792;
-	para->dx_odt = 0x07070707;
-	para->tpr6 = 0x4400c080;
-	para->tpr10 = 0x40ef6663;
-	para->tpr11 = 0x1a1a1a16;
-	para->tpr12 = 0x14121210;
-}
-
-static bool walnutpi_try(const char *tag, struct dram_para *para,
-			 struct dram_config *config)
-{
-	walnutpi_any_phy = false;
-	if (mctl_auto_detect_rank_width(para, config)) {
-		printf(" %s+", tag);
-		return true;
-	}
-	printf(" %s-%c", tag, walnutpi_any_phy ? 'w' : 'p');
-	return false;
+	para->tpr6 = CONFIG_DRAM_SUN50I_H616_TPR6;
+	para->tpr10 = CONFIG_DRAM_SUN50I_H616_TPR10;
+	para->tpr11 = CONFIG_DRAM_SUN50I_H616_TPR11;
+	para->tpr12 = CONFIG_DRAM_SUN50I_H616_TPR12;
 }
 
 unsigned long sunxi_dram_init(void)
@@ -1545,44 +1495,23 @@
 	setbits_le32(&prcm->res_cal_ctrl, BIT(8));
 	clrbits_le32(&prcm->ohms240, 0x3f);
 
-	/* 15:50:30 trained as DDR3; LPDDR4 TPR/bit-delay starves that SKU.
-	 * DDR3 first at 1.3 V (1B schematic), Zero2 timings; then LPDDR4 at 1.1 V.
+	/* Official AUTO: DCDC3 already 1.1 V. DDR3 four rank/width
+	 * tries, then LPDDR4. No write leveling.
 	 */
-	walnutpi_dram_volt(1300);
-	walnutpi_load_ddr3(&para, 720);
-	if (walnutpi_try("d3A", &para, &config))
-		goto detected;
-	walnutpi_load_ddr3(&para, 672);
-	if (walnutpi_try("d3B", &para, &config))
-		goto detected;
-	walnutpi_load_ddr3(&para, 792);
-	if (walnutpi_try("d3C", &para, &config))
-		goto detected;
-	walnutpi_load_ddr3_vendor(&para);
-	if (walnutpi_try("d3D", &para, &config))
+	walnutpi_para_ddr3(&para);
+	printf(" ddr3");
+	if (mctl_auto_detect_rank_width(&para, &config))
 		goto detected;
+	printf("-");
 
-	walnutpi_dram_volt(1100);
-	walnutpi_load_ddr3_vendor(&para);
-	if (walnutpi_try("d3E", &para, &config))
-		goto detected;
-	walnutpi_load_lpddr4(&para, 0x4400c080, 0x40ef6663, 0x24242323, 0x0e0e0e0e);
-	if (walnutpi_try("lp4A", &para, &config))
-		goto detected;
-	walnutpi_load_lpddr4(&para, 0x44000000, 0x402f6663, 0x24242624, 0x0f0f100f);
-	if (walnutpi_try("lp4B", &para, &config))
-		goto detected;
-	walnutpi_load_lpddr4(&para, 0x48808080, 0x402f6663, 0x26262524, 0x100f100f);
-	if (walnutpi_try("lp4C", &para, &config))
+	walnutpi_para_lpddr4(&para);
+	printf(" lp4");
+	if (mctl_auto_detect_rank_width(&para, &config))
 		goto detected;
+	printf("-");
 	panic("This DRAM setup is currently not supported.\n");
 
 detected:
-	if (!walnutpi_sdram_works()) {
-		printf(" dead");
-		panic("This DRAM setup is currently not supported.\n");
-	}
-
 	mctl_auto_detect_dram_size(&para, &config);
 	if (!mctl_core_init(&para, &config))
 		panic("DRAM core init failed.\n");
